### ISSUE TITLE
Export variables with hint float and infer type/wrong type don't use the inspector values

### ISSUE THREAD

**Issue by mrcdk** [godotengine/godot#21577]
**Godot version:**

Godot master https://github.com/godotengine/godot/commit/7b7f4b20cb6070b632824915033fca0d181b4c7f

**OS/device including version:**

Windows 10

**Issue description:**

![https://i.imgur.com/YSVyHH5.png](https://i.imgur.com/YSVyHH5.png)

- `something` has the hint `float` and it infers the type from the value `6`. Its type is `int` and won't use the inspector value
- `deacceleration` is the same but with the explicit type `int`. It won't use the inspector value either.
- `another` has hint `float` but no type. It uses the value of the inspector.
- `walk_speed` has hint `float` and infers type from `3.0`. Its type is `float` and uses the inspector value
- `run_speed` is the same but with explicit type. It uses the value of the inspector.

Maybe related closed issue https://github.com/godotengine/godot/issues/20402

cc @vnen

**Comment by akien-mga** [comment]
Seems still reproducible in the current master branch.

**Comment by Xrayez** [comment]
Hmm, stumbled upon this too, took me a few minutes to figure out this is a bug, though I imagine this could lead to hard-to-find bugs in developed projects if goes unnoticed.

Attaching minimal project:
[export-var-typed-hint.zip](https://github.com/godotengine/godot/files/3055869/export-var-typed-hint.zip)

**Comment by akien-mga** [comment]
The minimal project in https://github.com/godotengine/godot/issues/21577#issuecomment-480961828 doesn't make the problem very obvious to me, the value is declared with `:= 0` so it's strongly typed as integer, so both asserts fail (which might not be what you had initially? as the comment on the first assert says "ok").

Reproduction code for OP:
```
export (float) var something := 6
export (float) var deacceleration : int = 10
export (float) var another = 2
export (float) var walk_speed := 3.0
export (float) var run_speed : float = 5

func _ready():
	print("INT: ", TYPE_INT)
	print("FLOAT: ", TYPE_REAL)
	for i in [something, deacceleration, another, walk_speed, run_speed]:
		print(typeof(i), " ", i)
```

With default values (no change in inspector), it prints:
```
3 6
3 10
3 2
3 3
3 5
```

If I add `+ 0.5` to all export variables in the inspector (which works, even for those supposed to be statically typed as `int`), it gives:
```
2 6
2 10
3 2.5
3 3.5
3 5.5
```

The two variables which are meant to be strongly typed as `int` are now properly identified as such and truncated. But it's surprising that it didn't work before changing the values.

**Comment by Xrayez** [comment]
> The minimal project in #21577 (comment) doesn't make the problem very obvious to me, the value is declared with := 0 so it's strongly typed as integer, so both asserts fail (which might not be what you had initially? as the comment on the first assert says "ok").

Yeah, seems like GDScript got a patch in the meantime. The first assert statement used to pass in spite of the variable declared as strongly typed `int`, `export(float)` used to determine the type instead.

**Comment by KoBeWi** [comment]
Resolved in 4.0. The `export(type)` syntax is no longer available, so it's impossible to cause mismatch.
